<a href="https://colab.research.google.com/github/mohamedamr13/Machine-Learning-/blob/main/Apriori.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import mlxtend
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder


In [2]:
dataset = [['Skirt', 'Sneakers', 'Scarf', 'Pants', 'Hat'],

    ['Sunglasses', 'Skirt', 'Sneakers', 'Pants', 'Hat'],

    ['Dress', 'Sandals', 'Scarf', 'Pants', 'Heels'],

    ['Dress', 'Necklace', 'Earrings', 'Scarf', 'Hat', 'Heels', 'Hat'],

   ['Earrings', 'Skirt', 'Skirt', 'Scarf', 'Shirt', 'Pants']]

Preparing Data using one hot encoding 


In [3]:
te=TransactionEncoder()
te_ary=te.fit(dataset).transform(dataset)    #Apply one-hot-encoding on our dataset
df=pd.DataFrame(te_ary, columns=te.columns_)  #Creating a new DataFrame from our Numpy array
df


Dress  Earrings    Hat  Heels  ...  Shirt  Skirt  Sneakers  Sunglasses
0  False     False   True  False  ...  False   True      True       False
1  False     False   True  False  ...  False   True      True        True
2   True     False  False   True  ...  False  False     False       False
3   True      True   True   True  ...  False  False     False       False
4  False      True  False  False  ...   True   True     False       False

[5 rows x 12 columns]

Support 

In [4]:
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules 

frequent_itemsets=apriori(df, min_support=0.6, use_colnames=True) #Instead of column indices we can use column names.
frequent_itemsets

support        itemsets
0      0.6           (Hat)
1      0.8         (Pants)
2      0.8         (Scarf)
3      0.6         (Skirt)
4      0.6  (Scarf, Pants)
5      0.6  (Pants, Skirt)

Confidence 

In [5]:
association_rules(frequent_itemsets,metric="confidence",min_threshold=0.2) # associate itemsets with confidence over 20%.


antecedents consequents  antecedent support  ...    lift  leverage  conviction
0     (Scarf)     (Pants)                 0.8  ...  0.9375     -0.04         0.8
1     (Pants)     (Scarf)                 0.8  ...  0.9375     -0.04         0.8
2     (Pants)     (Skirt)                 0.8  ...  1.2500      0.12         1.6
3     (Skirt)     (Pants)                 0.6  ...  1.2500      0.12         inf

[4 rows x 9 columns]

In [ ]:
association_rules(frequent_itemsets,metric="lift",min_threshold=1.25)


antecedents consequents  antecedent support  consequent support  support  \
0     (Skirt)     (Pants)                 0.6                 0.8      0.6   

   confidence  lift  leverage  conviction  
0         1.0  1.25      0.12         inf

## Suggested Strategies 


- Have packages containing scarfs and pants 
- Put additional racks of scarfs next to pants and vice versa since they are both frequently purchased together
- Put additional racks of skirts next to pants since it is likely for one who bought pants to buy skirts 

### Applying Apriori on large database 

In [ ]:
import csv
datafile = open('Market_Basket_Optimisation.csv', 'r')
datareader = csv.reader(datafile, delimiter=',')
data = []
for row in datareader:
     data.append(row)  

In [ ]:
te=TransactionEncoder()
te_ary=te.fit(data).transform(data)    #Apply one-hot-encoding on our dataset
df=pd.DataFrame(te_ary, columns=te.columns_)  #Creating a new DataFrame from our Numpy array
df.isnull().sum().sum()


0

In [ ]:
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules
frequent_itemsets=apriori(df, min_support=0.03, use_colnames=True) #Instead of column indices we can use column names.
frequent_itemsets[ frequent_itemsets['support'] > 0.1 ]

support         itemsets
7   0.163845      (chocolate)
11  0.179709           (eggs)
13  0.170911   (french fries)
18  0.132116      (green tea)
24  0.129583           (milk)
25  0.238368  (mineral water)
31  0.174110      (spaghetti)

In [ ]:
association_rules(frequent_itemsets,metric="confidence",min_threshold=0.0000000000000000000001) 

antecedents          consequents  ...  leverage  conviction
0           (chocolate)               (eggs)  ...  0.003751    1.028711
1                (eggs)          (chocolate)  ...  0.003751    1.025603
2           (chocolate)       (french fries)  ...  0.006393    1.049383
3        (french fries)          (chocolate)  ...  0.006393    1.046827
4           (chocolate)               (milk)  ...  0.010898    1.082736
5                (milk)          (chocolate)  ...  0.010898    1.111823
6           (chocolate)      (mineral water)  ...  0.013604    1.122357
7       (mineral water)          (chocolate)  ...  0.013604    1.073256
8           (chocolate)          (spaghetti)  ...  0.010668    1.085581
9           (spaghetti)          (chocolate)  ...  0.010668    1.079070
10       (french fries)               (eggs)  ...  0.005681    1.042232
11               (eggs)       (french fries)  ...  0.005681    1.039640
12               (milk)               (eggs)  ...  0.007509    1.076009
13               (eggs)               (milk)  ...  0.007509    1.050423
14      (mineral water)               (eggs)  ...  0.008090    1.043158
15               (eggs)      (mineral water)  ...  0.008090    1.062815
16          (spaghetti)               (eggs)  ...  0.005239    1.038081
17               (eggs)          (spaghetti)  ...  0.005239    1.036592
18       (french fries)      (mineral water)  ... -0.007011    0.948894
19      (mineral water)       (french fries)  ... -0.007011    0.965741
20      (mineral water)  (frozen vegetables)  ...  0.013007    1.064189
21  (frozen vegetables)      (mineral water)  ...  0.013007    1.218270
22          (green tea)      (mineral water)  ... -0.000430    0.995748
23      (mineral water)          (green tea)  ... -0.000430    0.997927
24        (ground beef)      (mineral water)  ...  0.017507    1.305401
25      (mineral water)        (ground beef)  ...  0.017507    1.088672
26        (ground beef)          (spaghetti)  ...  0.022088    1.373997
27          (spaghetti)        (ground beef)  ...  0.022088    1.163716
28               (milk)      (mineral water)  ...  0.017105    1.209650
29      (mineral water)               (milk)  ...  0.017105    1.089850
30               (milk)          (spaghetti)  ...  0.012900    1.137061
31          (spaghetti)               (milk)  ...  0.012900    1.093043
32      (mineral water)           (pancakes)  ...  0.011071    1.054100
33           (pancakes)      (mineral water)  ...  0.011071    1.180529
34      (mineral water)          (spaghetti)  ...  0.018223    1.102008
35          (spaghetti)      (mineral water)  ...  0.018223    1.159314

[36 rows x 9 columns]

In [ ]:
analysis = association_rules(frequent_itemsets,metric="lift",min_threshold=1.2) 
analysis[ (analysis['support'] > 0.03) & (analysis['confidence'] > 0.2) & (analysis['lift'] > 1.5)]

antecedents      consequents  ...  leverage  conviction
3                (milk)      (chocolate)  ...  0.010898    1.111823
11  (frozen vegetables)  (mineral water)  ...  0.013007    1.218270
12        (ground beef)  (mineral water)  ...  0.017507    1.305401
14        (ground beef)      (spaghetti)  ...  0.022088    1.373997
15          (spaghetti)    (ground beef)  ...  0.022088    1.163716
16               (milk)  (mineral water)  ...  0.017105    1.209650
17      (mineral water)           (milk)  ...  0.017105    1.089850
18               (milk)      (spaghetti)  ...  0.012900    1.137061
19          (spaghetti)           (milk)  ...  0.012900    1.093043

[9 rows x 9 columns]

# Data Analysis 


**1) Near Cashier Items**
Since mineral water is the item with the greatest support of 23.8% , Put Mineral Water directly on the counter or relativly near to the cashier , and in seperate , clear blocks for the customer to notice 

Other items with support greater than 10% are also candiates to this strategy including : milk , choclate , spaghetti , eggs , and green Tea

**2) Duos and Kits:**
Form combinations of items that are frequently bought together , put them on the same compartment and have frequent discounts on them 
The Duos with the highest lifts and confidence are the candidates.


> Examples 
       


In [ ]:
analysis[ (analysis['support'] > 0.03) & (analysis['confidence'] > 0.2) & (analysis['lift'] > 1.5)]

antecedents      consequents  ...  leverage  conviction
3                (milk)      (chocolate)  ...  0.010898    1.111823
11  (frozen vegetables)  (mineral water)  ...  0.013007    1.218270
12        (ground beef)  (mineral water)  ...  0.017507    1.305401
14        (ground beef)      (spaghetti)  ...  0.022088    1.373997
15          (spaghetti)    (ground beef)  ...  0.022088    1.163716
16               (milk)  (mineral water)  ...  0.017105    1.209650
17      (mineral water)           (milk)  ...  0.017105    1.089850
18               (milk)      (spaghetti)  ...  0.012900    1.137061
19          (spaghetti)           (milk)  ...  0.012900    1.093043

[9 rows x 9 columns]

The most frequently bought item feature ( spaghetti and ground beef ) would suggest that you place each on right next to the others in their compartments as well as creating DUOs or packages containing both of them at a discounted price 